util/ranger/detacher.go

```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+//            []*valueInfo              the values of every columns in the returned new condtions
+//            error                     if error gernerated, return error
+func AddGcColumn4EqCond(sctx sessionctx.Context,
+	cols []*expression.Column,
+	accessesCond []expression.Expression,
+	columnValues []*valueInfo) ([]expression.Expression, error) {
+
+	expr := cols[0].VirtualExpr.Clone()
+	record := make([]types.Datum, len(columnValues)-1)
+
+	for i := 1; i < len(columnValues); i++ {
+		cv := columnValues[i]
+		if cv == nil {
+			break
+		}
+		record[i-1] = *cv.value
+	}
+
+	mutRow := chunk.MutRowFromDatums(record)
+	evaluated, err := expr.Eval(mutRow.ToRow())
+	if err != nil {
+		return accessesCond, err
+	}
+	vi := &valueInfo{false, &evaluated}
+	con := &expression.Constant{Value: evaluated, RetType: cols[0].RetType}
+	// make a tidb_shard() function, e.g. `tidb_shard(a) = 8`
+	cond := expression.NewFunctionInternal(sctx, ast.EQ, cols[0].RetType, cols[0], con)
+	if cond == nil {
+		return accessesCond, nil
+	}
+
+	accessesCond[0] = cond
+	columnValues[0] = vi
+	return accessesCond, nil
+}
+
+// AddExpr4EqAndInCondition add the `tidb_shard(x) = xxx` prefix
+// Add tidb_shard() for EQ and IN function. e.g. input condition is `WHERE a = 1`,
+// output condition is `WHERE tidb_shard(a) = 214 AND a = 1`. e.g. input condition
+// is `WHERE a IN (1, ,2 ,3)`, output condition is `WHERE (tidb_shard(a) = 214 AND a = 1)
+// OR (tidb_shard(a) = 143 AND a = 2) OR (tidb_shard(a) = 156 AND a = 3)`
+// @param[in] conditions  the original condition to be processed
+// @param[in] cols        the columns of shard index, such as [tidb_shard(a), a, ...]
+// @param[in] lengths     the length for every column of shard index
+// @retval - the new condition after adding tidb_shard() prefix
+func AddExpr4EqAndInCondition(sctx sessionctx.Context, conditions []expression.Expression,
+	cols []*expression.Column) ([]expression.Expression, error) {
+
+	accesses := make([]expression.Expression, len(cols))
+	newConditions := make([]expression.Expression, 0, len(conditions))
+	columnValues := make([]*valueInfo, len(cols))
+	offsets := make([]int, len(conditions))
+	addGcCond := true
+
+	for i, cond := range conditions {
+		offset := getPotentialEqOrInColOffset(sctx, cond, cols)
+		offsets[i] = offset
+		if offset == -1 {
+			continue
+		}
+		if accesses[offset] == nil {
+			accesses[offset] = cond
+			continue
+		}
+		// if the same field appear twice or more, don't add tidb_shard()
+		// e.g. `WHERE a > 100 and a < 200`
+		addGcCond = false
+	}
+
+	for i, cond := range accesses {
+		if cond == nil {
+			continue
+		}
+		if !allEqOrIn(cond) {
+			addGcCond = false
+			break
+		}
+		columnValues[i] = extractValueInfo(cond)
+	}
+
+	if !addGcCond || !NeedAddGcColumn4ShardIndex(cols, accesses, columnValues) {
+		return conditions, nil
+	}
+
+	// remove the accesses from newConditions
```

Comment:
Better to comment about the meaning of `accesses` and `newAccesses`.

Reply:
ok